common/src/app/common/pages/changes.cljc

```diff
@@ -54,45 +68,50 @@
-                             :id id)]
-              (if (and (or (nil? parent-id) (contains? objects parent-id))
-                       (or (nil? frame-id) (contains? objects frame-id)))
-                (-> data
-                    (update :objects assoc id obj)
-                    (update-in [:objects parent-id :shapes]
-                               (fn [shapes]
-                                 (let [shapes (or shapes [])]
-                                   (cond
-                                     (some #{id} shapes)
-                                     shapes
-
-                                     (nil? index)
-                                     (if (= :frame (:type obj))
-                                       (d/concat [id] shapes)
-                                       (conj shapes id))
-
-                                     :else
-                                     (cph/insert-at-index shapes index [id])))))
-
-                    (cond-> (and (:shape-ref (get-in data [:objects parent-id]))
-                                 (not= parent-id frame-id)
-                                 (not ignore-touched))
-                      (->
-                        (update-in [:objects parent-id :touched]
-                                   cph/set-touched-group :shapes-group)
-                        (d/dissoc-in [:objects parent-id :remote-synced?]))))
-                data)))]
+  [data {:keys [id obj page-id component-id frame-id parent-id index ignore-touched]}]
+  (letfn [(update-parent-shapes [shapes]
+            ;; Ensure that shapes is always a vector.
+            (let [shapes (into [] shapes)]
+              (cond
+                (some #{id} shapes)
+                shapes
+
+                (nil? index)
+                (if (= :frame (:type obj))
+                  (into [id] shapes)
+                  (conj shapes id))
+
+                :else
+                (cph/insert-at-index shapes index [id]))))
+
+          (update-parent [parent]
+            (-> parent
+                (update :shapes update-parent-shapes)
+                (update :shapes vec-without-nils)
+                (cond-> (and (:shape-ref parent)
+                             (not= (:id parent) frame-id)
+                             (not ignore-touched))
+                  (-> (update :touched cph/set-touched-group :shapes-group)
+                      (dissoc :remote-synced?)))))
+
+          (update-objects [objects parent-id]
+            (if (and (or (nil? parent-id) (contains? objects parent-id))
+                     (or (nil? frame-id) (contains? objects frame-id)))
+              (-> objects
+                  (assoc id (-> obj
+                                (assoc :frame-id frame-id)
+                                (assoc :parent-id parent-id)
+                                (assoc :id id)))
+                  (update parent-id update-parent))
+              objects))
+
+          (update-page [data]
```

Comment:
This should be `(update-container [container]`.

Container is the abstraction we made for structures that may be a page or a component.

Reply:
you're right